name: Ka0s YAML Lint

on:
  workflow_dispatch: # Dispara el workflow manualmente
    inputs:
      kaos-origin:
        description: 'Dispatcher Workflow Run ID'
        default: ""
        required: true
      kaos-files:
        description: 'Files Involved'
        default: ""
        required: true
      kaos-issue-id:
        description: 'Issue ID'
        default: "Kaos Issue ID"
        required: true
  

permissions:
  contents: write
  actions: write
  issues: write
  pull-requests: write

env:
  KAOS_MODULE: "[Ka0S] YAMLLINT"
  GH_TOKEN: ${{ secrets.KAOS_TOKEN }}
  KAOS_YAMLLINT_CONFIG: ${{ secrets.KAOS_YAMLLINT_CONFIG }}

jobs:
  yaml-lint-process:
    runs-on: kaos-runner-yamllint

    steps:
      - id: repo # Checkout the repository to get the information of the commit
        uses: actions/checkout@v4
        with:
          fetch-depth: 0
          token: ${{ secrets.KAOS_TOKEN }}
    
      - id: python
        uses: actions/setup-python@v5
        with:
          python-version: '3.x'
  
      - id: install-yamllint
        run: |
          python -m pip install yamllint

      - id: check-yaml
        run: |
            # Invocamos la primera comprobaci칩n de base, sino se supera el proceso no continua
            yamllint -c $KAOS_YAMLLINT_CONFIG "${{ github.event.inputs.kaos-files }}" >> core/results/yaml-lint-results-${{ github.run_id }}.txt || true
              # Check if exist errors in file
              if [ ! -s .core/results/yaml-lint-results-${{ github.run_id }}.txt ]; then
                # Todo ha salido bien, a침adimos el comentario a la issue y saltamos al siguiente paso
                KAOS_ISSUE_BODY="The yaml lint process ${{ github.event.inputs.kaos-origin }} launch the check ${{ github.run_id }} and its result has been SUCCESS"
                gh workflow run issue.yaml --ref ${{ github.ref }} -f kaos-issue-title="" -f kaos-issue-body="$KAOS_ISSUE_BODY" -f kaos-issue-actions=comment -f kaos-issue-tag="" -f kaos-issue-id=${{ github.event.inputs.kaos-issue-id }}
                # una vez finalizado este proceso, en funcion del modulo donde esta ubicado el fichero que se ha comprobado ejecutara un jobs u otro
                # No existen mas m칩dulos por detras procedemos a indicar que todo ha ido correcto u que dejamos generada la PR
                gh workflow run issue.yaml --ref ${{ github.ref }} -f kaos-issue-title="" -f kaos-issue-body="Ka0S approve create pull request for issue ${{ github.event.inputs.kaos-issue-id }}.Please review all files implicated in pull request and approve to merge" -f kaos-issue-actions=comment -f kaos-issue-tag="" -f kaos-issue-id=${{ github.event.inputs.kaos-issue-id }}                
              else
                # Algo no ha salido bien, a침adimos el comentario a la issue y saltamos al siguiente paso
                KAOS_ISSUE_BODY="The yaml lint process ${{ github.event.inputs.kaos-origin }} launch the check ${{ github.run_id }} and its result has been FAILED, please check core/results/yaml-lint-results-${{ github.run_id }}.txt"
                gh workflow run issue.yaml --ref ${{ github.ref }} -f kaos-issue-title="" -f kaos-issue-body="$KAOS_ISSUE_BODY" -f kaos-issue-actions=comment -f kaos-issue-tag="" -f kaos-issue-id=${{ github.event.inputs.kaos-issue-id }} -f kaos-issue-body-file=core/results/yaml-lint-results-${{ github.run_id }}.txt
                # una vez finalizado este proceso, en funcion del modulo donde esta ubicado el ficher que se ha comprobado ejecutara un jobs u otro 
              fi

  handle_failure:
    runs-on: 
      group: kaos-test
    needs: [yaml-lint-process]
    if: ${{ failure() }}
    steps:
      - id: check-execution
        env:
          GH_TOKEN: ${{ secrets.KAOS_TOKEN }}
        run: |
          echo "Error detected in ${{ env.KAOS_MODULE }}, creating issue..."
          KAOS_ISSUE_TITLE="${{ env.KAOS_MODULE }} Action Failed: ${{ github.workflow }}"
          KAOS_ISSUE_BODY=$(cat <<'EOF'
          An error occurred in the ${{ env.KAOS_MODULE }} Action workflow **${{ github.workflow }}**.
          **${{ env.KAOS_MODULE }}  Message:**
          \`\`\`
          **Please check the job logs for ${{ secrets.KAOS_ACTIONS_URL }}${{ github.run_id }}**
          \`\`\`
          **Workflow Name:** ${{ github.workflow }}
          **Job Name:** ${{ github.job }}
          **Run ID:** ${{ github.run_id }}
          **Run Number:** ${{ github.run_number }}
          **Repository:** ${{ github.repository }}
          **Branch:** ${{ github.ref }}
          **Commit:** ${{ github.sha }}
          Please investigate the issue.
          EOF
          )
          gh workflow run issue.yaml --ref ${{ github.ref }} -f kaos-issue-title="$KAOS_ISSUE_TITLE" -f kaos-issue-body="$KAOS_ISSUE_BODY" -f kaos-issue-actions=create -f kaos-issue-tag="bug"   
